### ISSUE TITLE
[Feature Request] Filter by "changed user data or updated app" and "last backup date"

### ISSUE THREAD

**Issue by dagaz** [machiav3lli/oandbackupx#241]
It would be great, if it would be possible to filter the app list by two new criteria:
1. by changed user data or updated app
2. by date of last backup

**Is your feature request related to a problem? Please describe.**
"changed user data or updated app": 
I would like to reduce backup time and write cycles. So it would be helpful if it would be possible to filter all apps where is a change of the app itself (update) OR a change of the user data (more important than the app itself ;)). Right now it can just be filtered by updated apps. But sometimes there is a big change for user data as well so the easiest way now is to backup all apps, regardless of changes.

"by date of last backup":
Sometimes I get errors for the backup process. I would like to filter the apps for which no backup was created so that I can manually redo a backup.

**Describe the primer solution**
I would like if there would be added the two filter options for the filter menu.
If the user data changed could be taken from the size of the apps user data. 
Backup dates could be generated as a list from the file dates.

**Describe alternative solutions**
I am not sure if there is an alternative. Regarding errors it might be possible to filter by "apps with backup errors" - but i think that could be the same effort.

**System Information(please complete the following information):**
 - Device: Oneplus 7 Pro
 - Android Version: 10
 - ROM: LineageOS
 - App's Version: 4.0.0 (F-Droid)

**Comment by machiav3lli** [comment]
1. This would posdibly never make it, as we had long discussions on other issues and in the chat group. The benifits are only marginal considering how much the data of the apps not stable compared to 3-4 years ago. So this doesn't come in question for now.

2. this option is already present in the main filters with the tag old backups. Including it in schedules modes shouldn't be a problem in future releases.

**Comment by dagaz** [comment]
Thank you for you quick answer.

>     1. This would posdibly never make it, as we had long discussions on other issues and in the chat group. The benifits are only marginal considering how much the data of the apps not stable compared to 3-4 years ago. So this doesn't come in question for now.

What a pity! But I understand that you have much more important priorities.
 
>     2. this option is already present in the main filters with the tag old backups. Including it in schedules modes shouldn't be a problem in future releases.

That would be nice.

**Comment by leso-kn** [comment]
Hello! Revisiting this issue in 2021 I'd like to add a comment on the first criteria described:
'only backup when app data was modified'

A scenario, where this could be beneficial is the case of backing up directly over a network (for example by mounting the target location [via rclone](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone#work-profiles--users)).

@machiav3lli I didn't quite understand your comment 'considering how much the data of the apps not stable compared to 3-4 years'. Could you ellaborate this point?

**Comment by leso-kn** [comment]
Thinking forward on this issue, there are two possible implementations that come to my mind,
maybe they're worth considering:

1. Determine modifications using fs (ext4) file modification time
2. Determine modifications using the Android Usage Statistics API

---

Ellaborating the two approaches:

**1. File modification time:**

- Check modification date of all files belonging to an app (recursively)
- Determine the latest (newest) modification time
- If no backup has been taken yet, perform a backup and remember the determined modification time (e.g. save to the OAndBackupX internal database)

Next time the backup is considered:

- Recalculate the modification time
- Compare if the new modification time is newer than the saved (remembered) one

**2. Android Usage Statistics API:**

The API provides methods to determine when an app was last used. This would be a great criteria to determine whether an app needs to be backed up or not, since (at least for launchable apps) it is very unlikely, that data changed if the app was not in use since the last backup.
Following the format from approach 1, the rough steps would be:

- When a backup is considered for an app:
- Use the `UsageStats` API to determine when the app was last used
- If no backup has been taken yet, perform a backup and remember the determined last-used time (e.g. save to the OAndBackupX internal database)

Next time the backup is considered:

- Use `UsageStats` API again, to determine the last-used time
- Compare if the new last-used time is later than the saved (remembered) time

**Comment by machiav3lli** [comment]
@leso-kn thanks for your documented research.

First more to what I meant by data of apps being not stable: background services to do some checks or syncing are being used more than before which is resulting some data files to change (mostly cache which is generally not needed). This dynamic behaviour makes the first solution obsolete.

Now to the suggested solutions:
1. Ignoring the forementioned point, such scanning using su cli will be a huge resources eater, which will reflect to a bad UX. The app should do the most possible with the least resources. An API is a far better - and maybe the only acceptable - solution.

2. This one seems like a great suggestion. Bit I would jump off comparing the two opened dates (before backup and last opened) and would just compared last opened with the last backup date. This should have the same usage.

The 2. with the change I mentioned can be a first iteration in an upcoming major release (unfortunately not 7.0.0, as it's already in rc-phase and should come next week).

**Comment by leso-kn** [comment]
Good idea with comparing it to the last backup date, right, that simplyfies the solution even more.

Looking forward to v7 next week! By the next major release you mean 8.0.0? It will probably take some time until that's being released, or does scheduling it for the major release mean it can be introduced in a 7.x.x version?

**Comment by machiav3lli** [comment]
Maybe I wasn't thinking that clear at the time. It could be possible to include such (after the change I suggested) in a minor release (7.1.0 should benext) as no change to the backup object is needed.

**Comment by leso-kn** [comment]
Sounds good. Is there a board/roadmap where this should be tracked as a potential feature for the next release?

**Comment by leso-kn** [comment]
@hg42 Sorry for the delay! Been quite busy lately.

Having a look at Android's UsageStats API ([here](https://developer.android.com/reference/android/app/usage/UsageStats#getLastTimeForegroundServiceUsed()) and [here](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/usage/UsageStats.java#453)), it seems to be possible to measure the last time a foreground service was used.

Since 'used' refers to whether the service is running in the background (regardless of changes to its data), this method might however not be very accurate for our case, so measuring changes in service-only apps might in fact be connected to more effort.

I believe having a changed-apps-only option, that only considers [Activity usage](https://developer.android.com/reference/android/app/usage/UsageStats#getLastTimeUsed()) for the moment, would already be a big advantage and could later be improved in order to support service-only apps as well.

What do you think?

**Comment by leso-kn** [comment]
@machiav3lli I see that a lot of feature requests have been submitted lately
(\> thus I was suggesting a step-by-step implementation for this one, to minimize the initial effort).

Is there a document (release schedule or the like), where potential features for upcoming versions are tracked?
Maybe picked by both priority and estimated effort.

**Comment by Victor239** [comment]
Came across this issue after finding out that one of my apps despite being used daily for important data hadn't been backed up for months. I was only using a "New and updated apps" filtered backup and didn't realise that since the app itself hadn't gotten any F-Droid updates it would never backup until then.

Any sort of "changed data" measurement would be good for me since otherwise I have to select all apps through an "old backups" filtered periodic backup anyway, plus I also use SuperFreezZ to stop background usage. Selecting only those which have used a foreground service however sounds like a more efficient direction to go in.

**Comment by cipitaua** [comment]
I also warlmy second this feature request, it's the same as TitaniumBackup's "redo backups for modified data", which is very useful when the backup folder is synced to the cloud (I use FolderSync for that and it's awesome).

**Comment by hg42** [comment]
the date/time of a directory isn't the same as the latest modification of any of the files contained in that directory (recursive!). This depends on the file system and the mount options. Usually changing a file does not change the directory.

```
➤ mkdir test      
➤ ls -dl test                              
drwxrwxr-x 2 harald harald 40 Nov 18 21:18 test
➤ touch test/a-file               
➤ ls -dl test      
drwxrwxr-x 2 harald harald 60 Nov 18 21:19 test

# directory was updated, because it's like a file of the entries and one entry was added

➤ mkdir test/a-dir  
➤ ls -dl test     
drwxrwxr-x 3 harald harald 80 Nov 18 21:20 test

# also changed the entries
                                                                                                                                                                          
➤ touch test/a-dir/another-file
➤ ls -dl test                  
drwxrwxr-x 3 harald harald 80 Nov 18 21:20 test

# created but in the same minute
                                                                                                                                                                          
➤ ls -dl test/a-dir            
drwxrwxr-x 2 harald harald 60 Nov 18 21:20 test/a-dir
➤ ls -dl test/a-dir/*
-rw-rw-r-- 1 harald harald 0 Nov 18 21:20 test/a-dir/another-file

# actually the directory was updated, but in the same minute
# (I did not wait long enough)
# ok, let's change the file again

➤ touch test/a-dir/another-file
➤ ls -dl test/a-dir/*          
-rw-rw-r-- 1 harald harald 0 Nov 18 21:21 test/a-dir/another-file

# ok, it's newer now

➤ ls -dl test/a-dir            
drwxrwxr-x 2 harald harald 60 Nov 18 21:20 test/a-dir

# but the directory did not change (q.e.d.)

➤ ls -dl test                  
drwxrwxr-x 3 harald harald 80 Nov 18 21:20 test

# and the top level directory didn't change either
```

**Comment by solf** [comment]
@hg42 thanks, you're right to call me out on this.

I didn't test it properly and then got stuck fixing a myriad of issues in #998 

But I did get around to fixing this :) New code https://github.com/solf/Neo-Backup/tree/7018f29f5e5d75e908ae54ecf7f47c62c9116581 here does directory scanning to a specified depth (3 by default) to detect changes.

It also does apk-dedup, so that the same apk isn't stored multiple times (it's all in effort to save backup space to be able to keep many weekly backup copies).

(There's also optimization attempt with hot-path scanning (first checking the place where the last change was detected) but it doesn't work now, I'm trying to fix it.)
